sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/AADOAuth2AuthorizedClientRepository.java

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends OAuth2AuthorizedClient> T loadAuthorizedClient(String clientRegistrationId,
+                                                                     Authentication principal,
+                                                                     HttpServletRequest request) {
+        if (this.isPrincipalAuthenticated(principal)) {
+            if (this.isWebAppClientMode || isAllInClientMode) {
+                OAuth2AuthorizedClient result = delegate.loadAuthorizedClient(clientRegistrationId, principal, request);
+                if (result != null || repo.isClientCredentials(clientRegistrationId)) {
+                    return (T) result;
+                }
+
+                if (repo.isAzureDelegatedClientRegistrations(clientRegistrationId)) {
+                    OAuth2AuthorizedClient azureClient = loadAuthorizedClient(getAzureClientId(), principal, request);
+                    if (azureClient == null) {
+                        throw new ClientAuthorizationRequiredException(AADClientRegistrationRepository.AZURE_CLIENT_REGISTRATION_ID);
+                    }
+                    OAuth2AuthorizedClient fakeAuthzClient = createFakeAuthzClient(azureClient, clientRegistrationId, principal);
+                    OAuth2AuthorizationContext.Builder contextBuilder =
+                        OAuth2AuthorizationContext.withAuthorizedClient(fakeAuthzClient);
+                    String[] scopes = null;
+                    if (!AADClientRegistrationRepository.isDefaultClient(clientRegistrationId)) {
+                        scopes = repo.findByRegistrationId(clientRegistrationId).getScopes().toArray(new String[0]);
+                    }
+                    OAuth2AuthorizationContext context = contextBuilder
+                        .principal(principal)
+                        .attributes(getAttributesConsumer(scopes))
+                        .build();
+                    OAuth2AuthorizedClient clientGotByRefreshToken = provider.authorize(context);
+                    try {
+                        ServletRequestAttributes attributes =
+                            (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
+                        delegate.saveAuthorizedClient(clientGotByRefreshToken, principal, request, attributes.getResponse());
+                    } catch (IllegalStateException exception) {
+                        LOGGER.warn("Can not save OAuth2AuthorizedClient.", exception);
+                    }
+                    return (T) clientGotByRefreshToken;
+                }
+            } else {
+                return service.loadAuthorizedClient(clientRegistrationId, principal.getName());
+            }
+            return null;
+        }
+        return this.anonymousAuthorizedClientRepository.loadAuthorizedClient(clientRegistrationId, principal, request);
```

Comment:
1. This will be triggered In which cause? 
2. Did you debug this? 
3. Could you please write integration test for this?
4. Can we combine anonymousAuthorizedClientRepository and delegate?

Reply:
This will be triggered In which cause?
-- The scenario Resource Server with Obo, if the request without the authorization header, the restricted resource uses the `@RegisteredOAuth2AuthorizedClient`, then before the `OAuth2AuthorizedClientRepository` will load the client info form `InMemoryOAuth2AuthorizedClientService`, no session is available to save or load.

Did you debug this?
-- Of course.

Could you please write integration test for this?
-- let me add them.
Can we combine anonymousAuthorizedClientRepository and delegate?
-- No, delegate process is used for WebAppOnly and WebAppAndResoureServer, anonymousAuthorizedClientRepository process is for ResourceServeWithObo.